rllib/core/rl_trainer/torch/torch_rl_trainer.py

```diff
@@ -1,10 +1,185 @@
+    def apply_gradients(self, gradients: ParamDictType) -> None:
+
+        # make sure the parameters do not carry gradients on their own
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+
+        # set the gradient of the parameters
+        for pid, grad in gradients.items():
+            self._params[pid].grad = grad
+
+        # for each optimizer call its step function with the gradients
+        for optim in self._optim_to_param:
+            optim.step()
+
+    @override(RLTrainer)
+    def build(self) -> None:
+        # TODO (Kourosh): How do we handle model parallism?
+        # TODO (Kourosh): Instead of using _TorchAccelerator, we should use the public
+        # api in ray.train but allow for session to be None without any errors raised.
+        if self._use_gpu:
+            self._device = _TorchAccelerator().get_device()
+        else:
+            self._device = torch.device("cpu")
+        super().build()
+
+    @override(RLTrainer)
+    def _make_distributed(self) -> MultiAgentRLModule:
+        module = self._make_module()
+
+        # if the module is a MultiAgentRLModule and nn.Module we can simply assume
+        # all the submodules are registered. Otherwise, we need to loop through
+        # each submodule and move it to the correct device.
+        # TODO (Kourosh): This can result in missing modules if the user does not
+        # register them in the MultiAgentRLModule. We should find a better way to
+        # handle this.
+        if isinstance(module, torch.nn.Module):
+            module.to(self._device)
+            module = DDPRLModuleWrapper(module)
+        else:
+            for key in module.keys():
+                module[key].to(self._device)
+                module[key] = DDPRLModuleWrapper(module[key])
+
+        return module
+
+    @override(RLTrainer)
+    def _convert_batch_type(self, batch: MultiAgentBatch):
+        batch = NestedDict(batch.policy_batches)
+        batch = NestedDict(
+            {
+                k: torch.as_tensor(v, dtype=torch.float32, device=self._device)
+                for k, v in batch.items()
+            }
+        )
+        return batch
+
+    @override(RLTrainer)
+    def do_distributed_update(self, batch: MultiAgentBatch) -> Mapping[str, Any]:
+        # in torch the distributed update is no different than the normal update
+        return self._update(batch)
+
+    @override(RLTrainer)
+    def get_param_ref(self, param: ParamType) -> Hashable:
+        return param
+
+    @override(RLTrainer)
+    def get_parameters(self, module: RLModule) -> Sequence[ParamType]:
+        return list(module.parameters())
+
+    @override(RLTrainer)
+    def get_optimizer_obj(
+        self, module: RLModule, optimizer_cls: Type[Optimizer]
+    ) -> Optimizer:
+        # TODO (Kourosh): the abstraction should take in optimizer_config as a
+        # parameter as well.
+        lr = self.optimizer_config.get("lr", 1e-3)
```

Comment:
Shouldn't we make the optimizer class part of the optimizer config?

Reply:
Otherwise, this signature and internal behavior of this method is quite confusing.

Reply:
Yes optimizer_config is yet to be designed. It's gonna be a dataclass that will include all the weird settings that you'd may want your optimizers to have. For now let's just go with simple configs like before.

Reply:
Ok, fair enough.